### PR TITLE
Define relevant RGMB metadata in generate instead of in constructor, …

### PR DESCRIPTION
**Pull request by shikhar413**
refs #23002

### CODE REVIEW

**Comment by shikhar413** [comment]
@GiudGiud this MR fixes some issues with passing meta data from PCCMG to the RGMB mesh generators by defining them at the point which they are initialized instead of assuming they are all initialized at the constructor stage. Hopefully shouldn't be too time consuming to review

File: modules/reactor/src/meshgenerators/AssemblyMeshGenerator.C [hunk 1]
```diff
@@ -458,6 +447,22 @@ AssemblyMeshGenerator::AssemblyMeshGenerator(const InputParameters & parameters)
...
+    _max_radius_meta = getMeshProperty<Real>("max_radius_meta", name() + "_pattern");
+  if (hasMeshProperty("background_intervals_meta", name() + "_pattern"))
+    _background_intervals_meta =
+        getMeshProperty<unsigned int>("background_intervals_meta", name() + "_pattern");
+  if (hasMeshProperty("node_id_background_meta", name() + "_pattern"))
+    _node_id_background_meta =
+        getMeshProperty<dof_id_type>("node_id_background_meta", name() + "_pattern");
+  if (hasMeshProperty("pattern_pitch_meta", name() + "_pattern"))
+    _pattern_pitch_meta = getMeshProperty<Real>("pattern_pitch_meta", name() + "_pattern");
+  if (hasMeshProperty("azimuthal_anlge_meta", name() + "_pattern"))
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/AssemblyMeshGenerator.C:462]:
```suggestion
  if (hasMeshProperty("azimuthal_angle_meta", name() + "_pattern"))
```

File: modules/reactor/include/meshgenerators/AssemblyMeshGenerator.h [hunk 2]
```diff
@@ -88,6 +88,21 @@ class AssemblyMeshGenerator : public ReactorGeometryMeshBuilderBase
...
+
+  /// MeshMetaData: maximum node id of the background region
+  dof_id_type & _node_id_background_meta;
+
+  /// MeshMetaData: pitch size to be used when stitching with assembly meshes
+  Real & _pattern_pitch_meta;
+
+  /// MeshMetaData: azimuthal angles of all nodes
+  std::vector<Real> & _azimuthal_angle_meta;
+
```
Comment (GiudGiud) [inline modules/reactor/include/meshgenerators/AssemblyMeshGenerator.h:105]: do they need to be class attributes?
You are not using them anywhere. 

While I agree it's been done a lot so far, maybe we should reconsider storing twice all metadata as attributes

File: modules/reactor/include/meshgenerators/AssemblyMeshGenerator.h [hunk 3]
```diff
@@ -88,6 +88,21 @@ class AssemblyMeshGenerator : public ReactorGeometryMeshBuilderBase
   ///The name of the assembly outer boundary, equal to the concatenation of "outer_assembly_" and the assembly-type ID
   std::string _assembly_boundary_name;
 
+  /// Metadata of the maximum radius of ring regions
```
Comment (GiudGiud) [inline modules/reactor/include/meshgenerators/AssemblyMeshGenerator.h:91]: what ring? the pins?

File: modules/reactor/include/meshgenerators/AssemblyMeshGenerator.h (hunk 2 above)
Comment (GiudGiud) [reply modules/reactor/include/meshgenerators/AssemblyMeshGenerator.h:105]: I m ticking now because some of these are a little odd. The max ring radius isnt typically what you care about for an assembly, it s very tied to a downstream application, separate from this object. Same for the azimuthal angles

File: modules/reactor/src/meshgenerators/PinMeshGenerator.C [hunk 4]
```diff
@@ -478,6 +469,22 @@ PinMeshGenerator::PinMeshGenerator(const InputParameters & parameters)
...
+    _max_radius_meta = getMeshProperty<Real>("max_radius_meta", name() + "_2D");
+  if (hasMeshProperty("background_intervals_meta", name() + "_2D"))
+    _background_intervals_meta =
+        getMeshProperty<unsigned int>("background_intervals_meta", name() + "_2D");
+  if (hasMeshProperty("node_id_background_meta", name() + "_2D"))
+    _node_id_background_meta =
+        getMeshProperty<dof_id_type>("node_id_background_meta", name() + "_2D");
+  if (hasMeshProperty("pattern_pitch_meta", name() + "_2D"))
+    _pattern_pitch_meta = getMeshProperty<Real>("pattern_pitch_meta", name() + "_2D");
+  if (hasMeshProperty("azimuthal_anlge_meta", name() + "_2D"))
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/PinMeshGenerator.C:484]:
```suggestion
  if (hasMeshProperty("azimuthal_angle_meta", name() + "_2D"))
```

**Review by GiudGiud — CHANGES_REQUESTED** [review]
would need a test dumping the meta data off of an existing test for example, to make sure it works fine now

File: modules/reactor/include/meshgenerators/AssemblyMeshGenerator.h (hunk 2 above)
Comment (shikhar413) [reply modules/reactor/include/meshgenerators/AssemblyMeshGenerator.h:105]: Looked into the underlying submeshgenerator code a bit more closely, the only metadata that actually needs to be transferred is max_radius_meta from PinMeshGenerator (defined in PCCMG) to AssemblyMeshGenerator (used by PatternedHexMeshGenerator), and pattern_pitch_meta from AssemblyMeshGenerator (defined in PatternedHexMeshGenerator) to CoreMeshGenerator (used by PatternedHexMeshGenerator) . The rest of the metadata isn't actually used anywhere for the specific workflow of Pin MG -> Assembly MG -> Core MG. However, these metadata need to be transferred explicitly in PinMeshGenerator and AssemblyMeshGenerator since there are intermediate mesh generators (such as BoundaryDeletionGenerator) between Pin MG, Assembly MG, and Core MG, an metadata is not transferred between these intermediate mesh generators.

With this in mind, is the preferred approach just to define class attirbutes for max_radius_meta in PinMG and pattern_pitch_meta in AssemblyMG and delete all the other class attributes that were defined in this PR?

Comment (GiudGiud) [reply modules/reactor/include/meshgenerators/AssemblyMeshGenerator.h:105]: why use a class attribute for these two (max_radius_meta, pattern_pitch_meta) even? Cant you retrieve them from the metadata after it s been declared?

I understand all these metadata need to be passed, I m not sure why they need to be stored in the class

@miaoyinb is working on a cleaner metadata forwarding for the future btw

Comment (shikhar413) [reply modules/reactor/include/meshgenerators/AssemblyMeshGenerator.h:105]: Currently ‘declareMeshProperty’ can’t be called outside of the constructor class, otherwise Moose throws an error. Therefore the workaround, which is a bit long-winded, is to store the reference to declareMeshProperty as a class variable and then update the class variable once the variable value is available in generate()

Comment (GiudGiud) [reply modules/reactor/include/meshgenerators/AssemblyMeshGenerator.h:105]: why dont we just add a getter that lets you modify the metadata or a setter

Comment (GiudGiud) [reply modules/reactor/include/meshgenerators/AssemblyMeshGenerator.h:105]: @permcody @roystgnr thoughts?

Comment (shikhar413) [reply modules/reactor/include/meshgenerators/AssemblyMeshGenerator.h:105]: Just wondering, what would the getter or setter look like? Feels like it would be very similar to declareMeshProperty without the constraint for restricting the call to constructors only

Comment (GiudGiud) [reply modules/reactor/include/meshgenerators/AssemblyMeshGenerator.h:105]: declareMeshProperty would stay in the constructor. 
However, the setter would:
- only work if the property has been declared already
- modify the data
There's already a getter. It's just const so it cant be used to modify

Comment (shikhar413) [reply modules/reactor/include/meshgenerators/AssemblyMeshGenerator.h:105]: @GiudGiud I am running into an issue with updating the code according to the suggestions. I am creating a function `setMeshProperty` that can be called after the constructor, assuming that `declareMeshProperty` has been called for the metadata of interest. However, I am running into an error in the function `MooseApp::registerRestartableData`, since it doesn't allow you to modify the metadata after it has been declared (see below in the `if data.declared` block):
image

Any ideas how I should bypass this? Not sure if I should be modifying `MooseApp::registerRestartableData`

Comment (GiudGiud) [reply modules/reactor/include/meshgenerators/AssemblyMeshGenerator.h:105]: I dont think you should call registerRestartableData again if it s been called in declareMeshProperty.

Are you doing this to get a reference to the restartable data?
Or to "save" the data in _restartable_data ?

Comment (shikhar413) [reply modules/reactor/include/meshgenerators/AssemblyMeshGenerator.h:105]: Thanks, I was calling `registerRestartableData` but now I am getting the reference to the restartable data directly from RestartableDataMap, can you check the `setMeshProperty` function now and let me know if this is what you had in mind? If so, I can address the remaining comments

File: framework/include/meshgenerators/MeshGenerator.h [hunk 5]
```diff
@@ -203,6 +210,31 @@ MeshGenerator::declareMeshProperty(const std::string & data_name, const T & init
   return data;
 }
 
+template <typename T>
+T &
+MeshGenerator::setMeshProperty(const std::string & data_name, const T & data_value)
+{
+  if (!_app.executingMeshGenerators())
+    mooseError(
+        "Updating mesh meta data cannot occur in the constructor of mesh generators");
```
Comment (GiudGiud) [inline framework/include/meshgenerators/MeshGenerator.h:219]: why?

File: framework/include/meshgenerators/MeshGenerator.h [hunk 6]
```diff
@@ -203,6 +210,31 @@ MeshGenerator::declareMeshProperty(const std::string & data_name, const T & init
...
+        "Mesh meta data must have been initialized before calling setMeshProperty");
+
+  std::string full_name =
+      std::string(MeshMetaDataInterface::SYSTEM) + "/" + name() + "/" + data_name;
+
+  // Get metadata reference from RestartableDataMap and update its value
+  auto & restartable_data_map = _app.getRestartableDataMap(MooseApp::MESH_META_DATA);
+  auto iter = restartable_data_map.find(full_name);
+  RestartableData<T> & restartable_data_ref = static_cast<RestartableData<T> &>(*iter->second.value);
+  T & data = restartable_data_ref.set();
```
Comment (GiudGiud) [inline framework/include/meshgenerators/MeshGenerator.h:232]: this works but it would be cleaner to move this code to a getRestartableData routine?

File: framework/include/meshgenerators/MeshGenerator.h (hunk 5 above)
Comment (shikhar413) [reply framework/include/meshgenerators/MeshGenerator.h:219]: I was thinking we impose this criteria to make it clear that setMeshGenerator should only be used outside of the constructor, otherwise if you want to update the value in the constructor just use declareMeshGenerator? If you don't think this is necessary, then I can remove this condition

Comment (GiudGiud) [reply framework/include/meshgenerators/MeshGenerator.h:219]: yeah let s remove. We dont really want to add a unit test to get that error tested as well

**Comment by shikhar413** [comment]
> would need a test dumping the meta data off of an existing test for example, to make sure it works fine now

@GiudGiud Added unit tests to test that meta data is properly being transferred and confirmed that this test fails if code changes from this PR are not applied

File: modules/reactor/include/meshgenerators/AssemblyMeshGenerator.h (hunk 2 above)
Comment (shikhar413) [reply modules/reactor/include/meshgenerators/AssemblyMeshGenerator.h]: Removed all these class variables now

File: framework/include/meshgenerators/MeshGenerator.h (hunk 6 above)
Comment (shikhar413) [reply framework/include/meshgenerators/MeshGenerator.h]: I created a function `getRestartableMetaData ` in MooseApp.C, since this class exposes RestartableData-related variables. Let me know if this function should be moved somewhere else instead.

**Comment by shikhar413** [comment]
@GiudGiud FYI all comments should be addressed now, PR is ready for re-review

File: framework/include/meshgenerators/MeshGenerator.h [hunk 7]
```diff
@@ -68,6 +68,13 @@ class MeshGenerator : public MooseObject, public MeshMetaDataInterface
   template <typename T>
   T & declareMeshProperty(const std::string & data_name, const T & init_value);
 
+  /**
+   * Method for updating attributes to the mesh meta-data store, which can be invoked after the
+   * mesh generator constructor has been called.
```
Comment (GiudGiud) [inline framework/include/meshgenerators/MeshGenerator.h:73]: I would rephrase this to say the property must already have been declared.

File: framework/src/base/MooseApp.C [hunk 8]
```diff
@@ -1709,6 +1709,20 @@ MooseApp::hasRestartableMetaData(const std::string & name,
   }
 }
 
+RestartableDataValue &
+MooseApp::getRestartableMetaData(const std::string & name,
+                                 const RestartableDataMapName & metaname) const
+{
+  // Get metadata reference from RestartableDataMap and return its value
```
Comment (GiudGiud) [inline framework/src/base/MooseApp.C:1716]:
```suggestion
  // Get metadata reference from RestartableDataMap and return a (non-const) reference to its value
```

File: modules/reactor/src/meshgenerators/PinMeshGenerator.C [hunk 9]
```diff
@@ -368,28 +368,25 @@ PinMeshGenerator::PinMeshGenerator(const InputParameters & parameters)
...
   if (hasMeshProperty("background_intervals_meta", name() + "_2D"))
     declareMeshProperty("background_intervals_meta",
                         getMeshProperty<unsigned int>("background_intervals_meta", name() + "_2D"));
   if (hasMeshProperty("node_id_background_meta", name() + "_2D"))
     declareMeshProperty("node_id_background_meta",
-                        getMeshProperty<unsigned int>("node_id_background_meta", name() + "_2D"));
+                        getMeshProperty<dof_id_type>("node_id_background_meta", name() + "_2D"));
   if (hasMeshProperty("pattern_pitch_meta", name() + "_2D"))
     declareMeshProperty("pattern_pitch_meta",
                         getMeshProperty<Real>("pattern_pitch_meta", name() + "_2D"));
```
Comment (GiudGiud) [inline modules/reactor/src/meshgenerators/PinMeshGenerator.C:389]: can you do a for loop same as the AssemblyMeshGenerator ?
for conciseness

**Review by GiudGiud — CHANGES_REQUESTED** [review]
@YaqiWang @permcody you come up in a git blame for mesh metadata.

Are you ok with this? eg modifying it after declaration or do you foresee issues ?

@shikhar413 
can you please check if other mesh generators can make use of this to remove mesh metadata from the class attributes

I believe @miaoyinb is working on mesh metadata forwarding in another PR btw.

File: modules/reactor/src/meshgenerators/PinMeshGenerator.C (hunk 9 above)
Comment (shikhar413) [reply modules/reactor/src/meshgenerators/PinMeshGenerator.C:389]: Is there a way to pass datatypes (Real, unsigned int, etc.) in the loop?

Comment (GiudGiud) [reply modules/reactor/src/meshgenerators/PinMeshGenerator.C:389]: You’re right, might be tricky to know at compile time.

**Comment by permcody** [comment]
As per usual, I don't have a ton of time to dive into this to understand the implications. However, since I was tagged, let me dump some thoughts and see if they line up. Maybe if my intent is understood we can understand if these changes line up with that vision.

The point of the `MeshMetaData` structure was to create a general data structure that could be used to store any user-defined attribute primary so that our input files that contain MeshGenerators work properly in a recover scenario. The initial driver of this work was to find a way to keep simulations running that relayed upon querying the Mesh (old system) for attribute data that was used in other parts of the system (e.g. a UserObject that needed geometric information that could trivially be saved from the mesh system but would otherwise be hard to compute from just the mesh structure such as "pellet stack height", "fuel gap width", etc.) This was the driver for the system.

I created an interface for interacting with that system in a read-only mode. I couldn't think of any "good" reason for why anything other than the Mesh system needed to write those attributes. Thus the interface only gives the developers read-only access. The interface was initially added to just one ore two of the base systems in MOOSE where attributes were already be accessed. On recover, these attributes could still be accessed, even though the MeshGenerators are not created or executed.

@GiudGiud - You ask if there are any issues/concerns with modifying these parameters after they are declared. The short answer is no. However, I specifically did not intend for this to be a "writable global data structure" from everywhere. I think from a design perspective, "readable from everywhere is fine" and "writable from select, controlled systems, and periods of time would be a much safer design". Again, without diving into this PR, are we still comfortable that we haven't pushed the intent of this system too far? Let's discuss.

**Comment by permcody** [comment]
Actually looking at this a bit more.... Maybe I should have done this first before my word dump above. I don't see any issue with "setting" the values in "generate" versus in the constructor. In fact this is explicitly the intent! I don't think any of the trivial MeshGenerators in MOOSE required that capability, but it is definitely ok!

File: framework/src/base/MooseApp.C [hunk 10]
```diff
@@ -1709,6 +1709,20 @@ MooseApp::hasRestartableMetaData(const std::string & name,
   }
 }
 
+RestartableDataValue &
+MooseApp::getRestartableMetaData(const std::string & name,
+                                 const RestartableDataMapName & metaname) const
+{
```
Comment (GiudGiud) [inline framework/src/base/MooseApp.C:1715]: so if this is queried in a threaded region, it will fail because the map is not well protected. Let's add a thread id argument and add an error just like in registerRestartableData

File: framework/include/meshgenerators/MeshGenerator.h [hunk 11]
```diff
@@ -203,6 +210,21 @@ MeshGenerator::declareMeshProperty(const std::string & data_name, const T & init
   return data;
 }
 
+template <typename T>
+T &
+MeshGenerator::setMeshProperty(const std::string & data_name, const T & data_value)
+{
```
Comment (GiudGiud) [inline framework/include/meshgenerators/MeshGenerator.h:216]: based on what Cody mentioned, you were right adding a limitiation on when this should be run.

Let's allow it in the generate routine only (as you had iirc) since it's likely to error if objects are constructed in the wrong order and this relies on a previous object having declared the routine

**Review by GiudGiud — CHANGES_REQUESTED** [review]
Thanks for your thoughts @permcody

File: framework/include/meshgenerators/MeshGenerator.h (hunk 11 above)
Comment (shikhar413) [reply framework/include/meshgenerators/MeshGenerator.h:216]: Added the error condition back

File: framework/src/base/MooseApp.C (hunk 10 above)
Comment (shikhar413) [reply framework/src/base/MooseApp.C:1716]: Added a thread ID argument and error condition

**Comment by shikhar413** [comment]
> @shikhar413
can you please check if other mesh generators can make use of this to remove mesh metadata from the class attributes

@GiudGiud removed the mesh metadata from many of the existing cases where they are stored as class attributes. Requesting a review from @miaoyinb to see if these changes are suitable since these code changes touch a lot of his code.

There were cases where mesh metadata were set by reference in the following manner, which I have left as is, since it seemed like it would make sense to store the meta data references as class attributes in these instances:

In the MG constructor:
```
_meta_data_ref = declareMeshProperty("meta_data", getParam<>("meta_data"))
```

and then in `generate()`:
```
callExternalFunction(_meta_data_ref, other_arguments) // _meta_data_ref is set in this function
```
```

File: framework/src/base/MooseApp.C [hunk 12]
```diff
@@ -1709,6 +1709,25 @@ MooseApp::hasRestartableMetaData(const std::string & name,
   }
 }
 
+RestartableDataValue &
+MooseApp::getRestartableMetaData(const std::string & name,
+                                 const RestartableDataMapName & metaname,
+                                 THREAD_ID tid) const
+{
+  if (!metaname.empty() && tid != 0)
```
Comment (GiudGiud) [inline framework/src/base/MooseApp.C:1717]: why !metaname.empty() ?
If there s no name dont we just get a default map? 
not super familiar

File: framework/include/meshgenerators/MeshGenerator.h [hunk 13]
```diff
@@ -203,6 +210,24 @@ MeshGenerator::declareMeshProperty(const std::string & data_name, const T & init
...
+  if (!_app.executingMeshGenerators())
+    mooseError("Updating mesh meta data cannot occur in the constructor of mesh generators");
+
+  std::string full_name =
+      std::string(MeshMetaDataInterface::SYSTEM) + "/" + name() + "/" + data_name;
+
+  auto & restartable_data_ref = static_cast<RestartableData<T> &>(
+      _app.getRestartableMetaData(full_name, MooseApp::MESH_META_DATA, 0));
```
Comment (GiudGiud) [inline framework/include/meshgenerators/MeshGenerator.h:224]: can we get the actual thread id here?

Comment (GiudGiud) [reply framework/include/meshgenerators/MeshGenerator.h:224]: ok seems like not, MGs arent threaded

**Review by GiudGiud — COMMENTED** [review]
just this one comment to address. 
@miaoyinb please let us know what do you think of the change from class attributes to all metadata in these classes

File: framework/src/base/MooseApp.C (hunk 12 above)
Comment (shikhar413) [reply framework/src/base/MooseApp.C]: I took this directly from `registerRestartableData`. However, looking at the logic flow, the following lines should return an error if metaname corresponds to an empty map since it will try to look for `name` within the `metaname` map and hit the error message. 

```
  // Get metadata reference from RestartableDataMap and return a (non-const) reference to its value
  auto & restartable_data_map = getRestartableDataMap(metaname);
  auto iter = restartable_data_map.find(name);
  if (iter == restartable_data_map.end())
    mooseError("Unable to find RestartableDataValue object with name " + name +
               " in RestartableDataMap");
```

Removed that extra condition in the if statement since it not needed

**Comment by miaoyinb** [comment]
I will take a look, thanks. @shikhar413 @GiudGiud

File: framework/include/meshgenerators/MeshGenerator.h [hunk 14]
```diff
@@ -68,6 +68,13 @@ class MeshGenerator : public MooseObject, public MeshMetaDataInterface
   template <typename T>
   T & declareMeshProperty(const std::string & data_name, const T & init_value);
 
+  /**
+   * Method for updating attributes to the mesh meta-data store, which can only be invoked outside
+   * of the MeshGenerator constructor after the mesh generator property has already been declared.
```
Comment (GiudGiud) [inline framework/include/meshgenerators/MeshGenerator.h:73]:
```suggestion
   * Method for updating attributes to the mesh meta-data store, which can only be invoked in
   * the MeshGenerator generate routine only if the mesh property has already been declared.
```

**Comment by GiudGiud** [comment]
This might be controversial @shikhar413 but at this point, can we even make MeshGenerator not inherit from MeshMetaDataInterface ?

**Comment by shikhar413** [comment]
> This might be controversial @shikhar413 but at this point, can we even make MeshGenerator not inherit from MeshMetaDataInterface ?

Not sure what is meant exactly by this request but I tried removing the class inheritance of `MeshMetaDataInterface` from `MeshGenerator` and there were many errors complaining that getMeshProperty and hasMeshProperty (currently defined in MeshMetaDataInterface) could not be found. Let me know if I misunderstood what you were trying to say

**Comment by GiudGiud** [comment]
no I meant that. 

ok this is fine

File: modules/reactor/src/meshgenerators/AzimuthalBlockSplitGenerator.C [hunk 15]
```diff
@@ -102,8 +102,9 @@ AzimuthalBlockSplitGenerator::generate()
   MeshTools::Modification::rotate(mesh, 90.0, 0.0, 0.0);
   _azimuthal_angle_meta = azimuthalAnglesCollector(mesh, -180.0, 180.0, ANGLE_DEGREE);
   // So that this class works for both derived classes of PolygonMeshGeneratorBase
-  _pattern_pitch_meta = std::max(getMeshProperty<Real>("pitch_meta", _input_name),
-                                 getMeshProperty<Real>("pattern_pitch_meta", _input_name));
+  auto pattern_pitch_meta = std::max(getMeshProperty<Real>("pitch_meta", _input_name),
+                                     getMeshProperty<Real>("pattern_pitch_meta", _input_name));
+  setMeshProperty("pattern_pitch_meta", pattern_pitch_meta);
```
Comment (miaoyinb) [inline modules/reactor/src/meshgenerators/AzimuthalBlockSplitGenerator.C:107]: It generally looks good to me. Thank you @shikhar413 

Just a comment. It does not occur in this commit. But it is possible that `data_value` of `setMeshProperty()` may have a different type from the originally declared mesh metadata type. So it may be safer to specify the type when calling `setMeshProperty()`

Comment (GiudGiud) [reply modules/reactor/src/meshgenerators/AzimuthalBlockSplitGenerator.C:107]: That's a good idea. An alternative would be to error if the name is already used with a different data type

Comment (shikhar413) [reply modules/reactor/src/meshgenerators/AzimuthalBlockSplitGenerator.C:107]: I did confirm that the user can call `setMeshProperty` by defining `data_value` as an arbitrary type that is different from the originally intended data type and it compiled without error. I do like the idea of adding an error within the function itself as a way to safeguard against a user defining the incorrect datatype from outside. 

@GiudGiud is there a way to modify setMeshProperty to get the original datatype of the restartable data? Right now a `static_cast` is used within setMeshProperty which basically does the datatype conversion at runtime, not at compile time. Changing to `dynamic_cast` didn't fix the issue either, since I was able to define a meta data that was supposed to be Real to std::string

Comment (GiudGiud) [reply modules/reactor/src/meshgenerators/AzimuthalBlockSplitGenerator.C:107]: There's a prettyCppType routine you can look at how it works to see if we have something, in that same file maybe.

Otherwise one classic way is to check whether a dynamic cast of the pointer passes or fails.

Comment (shikhar413) [reply modules/reactor/src/meshgenerators/AzimuthalBlockSplitGenerator.C:107]: So the RestartableData class has a function type() that returns the type of the metadata using prettyCppType, I could have a condition like this within setMeshProperty:

```if (_app.getRestartableMetaData(full_name, MooseApp::MESH_META_DATA, 0).type() != MooseUtils::prettyCppType(&data_value)))
  // Throw error
```

Is that the preferred route? This basically enforces that the same datatype used to define the metadata must be used to redefine it, and conversions from long to int etc. probably won't be allowed since we are comparing to the exact name of the datatype

Comment (GiudGiud) [reply modules/reactor/src/meshgenerators/AzimuthalBlockSplitGenerator.C:107]: I think it s fair. Too bad for long-int conversions. It s risky anyway so I d rather we don’t support them

Comment (shikhar413) [reply modules/reactor/src/meshgenerators/AzimuthalBlockSplitGenerator.C:107]: Check has been added

**Comment by shikhar413** [comment]
@GiudGiud do let me know if there are any remaining comments to address, thank you

**Review by GiudGiud — APPROVED** [review]

**Comment by GiudGiud** [comment]
Sorry for the delay I've been on travel.
Merging, thank you for the fix and the additional capability permitting some clean up. This is a smart improvement to the system in terms of flexibility, at the framework level.
